---
title: "assignment8"
author: "ChiaraOrlacchio"
date: "2024-04-29"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---


##Download (a subset of) ATAC-seq peak counts in the hippocampus upon
stress (already in SummarizedExperiment format): #○
<https://ethz-ins.org/content/mouse_mm38_hippocampus.peakCounts.SE.rds>
#○(the data is from the mouse ensembl GRCm38 genome – you should already
have the genome, e.g. from week 6)

##Using this object, perform a chromVAR motif analysis, and run 2
differential motif accessibility analyses, respectively: #○ comparing
stressed (denoted ‘FSS’ – forced swim stress) and control animals #○
comparing male and female animals

##For each analysis, report the top most significant motifs, plot a
heatmap of the normalized accessibility scores across the samples for
those motifs, and write a short paragraph interpreting the results.

```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(motifmatchr)
  library(MotifDb)
  library(AnnotationHub)
  library(universalmotif)
  library(ggplot2)
  library(SummarizedExperiment) # data structure
  library(sechm) # for plotting heatmaps from a SummrizedExperiment
  library(BiocParallel) # for multithreading
  library(chromVAR) # for motif accessibility estimation
  library(limma) # for statistical analysis
})
#install.packages("Matrix", version="1.6.3")
# to control multithreading, unix users can use:
register(MulticoreParam(4))
```

## Download the data

```{r, eval=FALSE}
options(timeout=6000)
dir.create("peak")
download.file("https://ethz-ins.org/content/mouse_mm38_hippocampus.peakCounts.SE.rds", destfile = "peak/ATAC_Peaks.rds", mode = "wb")
peak <- readRDS("peak/ATAC_Peaks.rds")

# get mouse genome
dir.create("genome")
ah <- AnnotationHub()
genome <- ah[["AH68356"]]
# we'll load it into memory:
genome_seqs <- import(genome)
Biostrings::writeXStringSet(genome_seqs, "genome.fa")
# preparing the genome sequence file (here just using a subset)
genome_fa <- Rsamtools::FaFile("genome.fa")
```

## Get the database of motifs

```{r}
motifs <- MotifDb::query(MotifDb, "Mus")
# convert to a format motifmatchr can use, and use the gene symbols as names
motifs <- do.call(TFBSTools::PWMatrixList, setNames(universalmotif::convert_motifs(motifs, class="TFBSTools-PWMatrix"), mcols(motifs)$geneSymbol)) #setname -> we are gona name the motifs simply by their geneSymbol
motifs
```

#1. get the fragment counts for each peak in each sample --\> and add GC bias

```{r}
se <- chromVAR::addGCBias(peak, genome=genome_fa)
rowData(se)
```

#2. get the occurence of each motif in each peak

```{r}
# we find which peaks contain which motifs
moi <- motifmatchr::matchMotifs(motifs, subject=se, genome=genome_fa) #what peak contains what motif -> not about where only if it contains one yes/no? what peaks match which motifs
moi

assay(se) <- as.matrix(assay(se))
# for each peak, we identify similar peaks as background

bg <- chromVAR::getBackgroundPeaks(object=as.matrix(assays(se)$counts), niterations=1000, bias=rowData(se)$bias)
# for each motif, we computed per-sample deviations relative to the background
dev <- chromVAR::computeDeviations(object =as.matrix(assays(se)$counts), annotations=as.matrix(assay(moi)), background_peaks=bg) #each of the rows is a motif -> relative score for acessability
head(assay(dev))


colData(dev)$condition <- c("CTRL", "CTRL", "CTRL", "CTRL", "CTRL", "CTRL", "FSS", "FSS", "FSS", "FSS", "FSS", "FSS")
colData(dev)$sex <- c("Female", "Female", "Female", "Male", "Male", "Male")
colData(dev) 
```

```{r}
head(assays(dev)$z)
```

# Differential analysis -\> to analyse which of the motif shows a difference between the two groups (dex vs ctrl)

#3. use 1 and 2 to obtain aggregated motif score per sample

```{r}
dev$condition
dev$sex

mm <- model.matrix(~sex+condition, data=as.data.frame(colData(dev))) 
mm #so we can see the name of the coefficients
```

```{r}
fit <- eBayes(lmFit(assays(dev)$z, mm)) #limma will do statistics and we will pass the z-scores (=activity score/acessability score); here we only have one coefficient -> the conditions
```

#FSS vs CTRL
```{r}
#FSS vs CTRL
res_stress <- as.data.frame(limma::topTable(fit, coef="conditionFSS", number = Inf)) #specify which coefficient we want to test
head(res_stress)
dim(res_stress)
res_stress$TF <- row.names(res_stress)
res_stress <- res_stress[!duplicated(res_stress$ID), ]#gets rid of duplicates
row.names(res_stress) <- res_stress$ID

ggplot(res_stress, aes(logFC, -log10(adj.P.Val), label=TF)) + geom_text() 
library(dplyr)
significant_motifs <- res_stress %>%
  filter(adj.P.Val < 0.05) %>%   # Filtering for significance
  arrange(adj.P.Val) 
head(significant_motifs, 6) #The most significant motifs -> significant difference of these motifs in stressed animals compared to the same ones in controls. And based on logFC (which is positive for all the top 10) the accessabilty of the motif in FSS animals is higher compared to control animals

sechm(dev, features = head(row.names(res_stress)), assayName="z", breaks = 1) #gives us z-score of significant motifs for the two different conditions: CTRL and FSS --> the stressed animals show a higher z-score -> higher acessability for stressed animals in males

```

#Male vs Female
```{r}
#Male vs Female
res_sex <- as.data.frame(limma::topTable(fit, coef="sexMale", number = Inf)) #specify which coefficient we want to test
head(res_sex)
dim(res_sex)
res_sex$TF <- row.names(res_sex)
res_sex <- res_sex[!duplicated(res_sex$ID), ]#gets rid of duplicates
row.names(res_sex) <- res_sex$ID

ggplot(res_sex, aes(logFC, -log10(adj.P.Val), label=TF)) + geom_text() 
# Filtering for significance
library(dplyr)
significant_motifs <- res_sex %>%
  filter(adj.P.Val < 0.05) %>%  
  arrange(adj.P.Val) 
head(significant_motifs, 6) #The most significant motifs -> significant difference of these motifs in males compared to the same ones in females. And based on logFC (which is negative for all the top 10) the accessabilty of the motif in males is lower compared to females

sex_annotation <- colData(dev)$sex  # 'sex' should be a column indicating 'male' or 'female'

# Prepare the annotation as a factor to color code it
sex_annotation_factor <- factor(sex_annotation, levels = c("male", "female"))
col_annotation <- HeatmapAnnotation(
    sex = sex_annotation_factor,
    col = list(sex = c(male = "orange", female = "green"))  # You can choose your colors
)
sechm(dev, features = head(row.names(res_sex)), assayName="z", top_annotation = col_annotation, breaks = 1) #gives us z-score of different motifs for male vs female and for the two conditions (CTRL - FSS) --> males show a higher z-score than females --> higher acessability for the listed motifs in males
```

# Interpretation

## CTRL - FSS:

-VolcanoPlot:The 6 most significant motifs --> significant difference of
these motifs in stressed animals compared to the same ones in controls.
And based on logFC (which is positive for all the top 10) the
accessabilty of the motif in FSS animals is higher compared to control
animals.

-Heatmap: gives us z-score of significant motifs for the two different
conditions: CTRL and FSS --> the stressed animals show a higher z-score --> higher acessability for the listed motifs in stressed animals; Only one column from the stressed condition shows no difference in the z-score for any of the motifs (my interpretation: this one columne is one animal that did not behave as expected)

## Female - Male: 

-VolcanoPlot: #The 6 most significant motifs -> significant difference of these motifs in males compared to the same ones in females. And based on logFC (which is negative for all the top
10) the accessabilty of the motif in males is lower compared to females

-Heatmap: #gives us z-score of different motifs for male vs female and
for the two conditions (CTRL - FSS) --> males show a higher z-score than females --> higher acessability for the listed motifs in males
